File: android/src/main/java/com/pingidentity/credentials/wallet/PingOneCredentialsSDK.java
package com.pingidentity.credentials.wallet;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.pingidentity.sdk.pingonewallet.types.regions.PingOneRegion;

public class PingOneCredentialsSDK extends ReactContextBaseJavaModule {

    private static final String ID_ROUTING_SERVICE_URL = "https://api.pingone.com/v1/";

    private PingOneWalletHelper pingOneWalletHelper = null;
    private final ReactApplicationContext reactApplicationContext;

    PingOneCredentialsSDK(ReactApplicationContext context) {
       super(context);
       this.reactApplicationContext = context;
   }

    @NonNull
    @Override
    public String getName() {
        return "PingOneCredentialsSDK";
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    public void initializeSDK(String region, Promise promise) {
        try {
            PingOneRegion pingOneRegion = getPingOneRegion(region);

            // in case initializeSDK is called multiple times.
            if (this.pingOneWalletHelper != null) {
                promise.resolve(pingOneWalletHelper.getApplicationInstanceId(pingOneRegion));
                return;
            }

            PingOneWalletHelper.initializeWallet(this.reactApplicationContext, pingOneRegion, helper -> {
                this.pingOneWalletHelper = helper;
                promise.resolve(helper.getApplicationInstanceId(pingOneRegion));
            }, throwable -> {
                promise.reject("Error initializing Wallet", throwable);
            });
        } catch (Exception e) {
            promise.reject("Error initializing Wallet", e);
        }
    }


    private PingOneRegion getPingOneRegion(String region) {
        return switch (region.toLowerCase()) {
            case "na" -> PingOneRegion.NA;
            case "eu" -> PingOneRegion.EU;
            case "ca" -> PingOneRegion.CA;
            case "apac" -> PingOneRegion.APAC;
            case "apac_2" -> PingOneRegion.APAC_2;
            default -> throw new RuntimeException("region " + region + " not supported");
        };
    }

    @ReactMethod(isBlockingSynchronousMethod = false)
    public void processRequest(String request, Promise promise) {
       if (this.pingOneWalletHelper == null) {
           promise.reject("PingOneWallet not initialized", (Throwable) null);
           return;
       }
       pingOneWalletHelper.processPingOneRequest(request, promise);
    }

    @ReactMethod(isBlockingSynchronousMethod = false)
    public void checkForMessages(Promise promise) {
        if (this.pingOneWalletHelper == null) {
            promise.reject("PingOneWallet not initialized", (Throwable) null);
            return;
        }
        pingOneWalletHelper.checkForMessages();
        promise.resolve("Message checking initiated");
    }

    @ReactMethod(isBlockingSynchronousMethod = false)
    public void getCredentialsList(Promise promise) {
        if (this.pingOneWalletHelper == null) {
            promise.reject("PingOneWallet not initialized", (Throwable) null);
            return;
        }
        pingOneWalletHelper.getCredentialsList(promise);
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    public void deleteCredential(String credentialId) {
       this.pingOneWalletHelper.deleteCredential(credentialId);
    }

    @ReactMethod(isBlockingSynchronousMethod = false)
    public void readNotifications(Promise promise) {
        this.pingOneWalletHelper.readNotifications(promise);
    }
}

File: ios/PingOneCredentialsSDK.swift
//
//  PingOneCredentialsSDK.swift
//  RNPingOneCredentialsSDK
//
//  Created by Gaurav Khot on 3/19/25.
//

import Foundation
import React
import PingOneWallet

@objc(PingOneCredentialsSDK)
class PingOneCredentialsSDK: RCTEventEmitter {
    
    private var pingOneWalletHelper: PingOneWalletHelper?
    
    
    override func supportedEvents() -> [String]! {
            return [
                "P1SDK_handleCredentialIssuance",
                "P1SDK_handleCredentialRevocation",
                "P1SDK_handlePairingRequest",
                "P1SDK_presentCredential",
                "P1SDK_handlePresentationAction",
                "P1SDK_handlePairingEvent",
                "P1SDK_handleErrorEvent"
            ]
        }
        
        private func emitEvent(eventName: String, map: [String: Any]) {
            if self.bridge != nil {
                self.sendEvent(withName: eventName, body: map)
            }
        }
    
    @objc(initializeSDK:resolver:rejecter:)
    func initializeSDK(_ region: String, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject:@escaping RCTPromiseRejectBlock) -> Void {
        guard let pingOneRegion = self.getPingOneRegion(region) else {
            reject("error", "region \(region) not supported", nil)
            return
        }
        
        if let pingOneWalletHelper = self.pingOneWalletHelper,
           let applicationInstanceId = pingOneWalletHelper.getApplicationInstanceId(region: pingOneRegion){
            resolve(applicationInstanceId)
            return
        }
        
        PingOneWalletClient.Builder(forRegion: pingOneRegion).build() // TODO: take region parameter
            .onError { error in
                reject("error", error.localizedDescription, error)
            }
            .onResult { client in
                self.pingOneWalletHelper = PingOneWalletHelper(client)
                self.pingOneWalletHelper?.eventCallback = { [weak self] eventName, payload in
                    self?.emitEvent(eventName: eventName, map: payload)
                }
                if let applicationInstanceId = client.getApplicationInstance(forRegion: pingOneRegion)?.getId() {
                    print("*********", applicationInstanceId)
                    resolve(applicationInstanceId)
                }
            }
    }
    
    private func getPingOneRegion(_ region: String) -> PingOneRegion? {
        switch region.lowercased() {
        case "na" : return .NA
        case "eu" : return .EU
        case "apac" : return .APAC
        case "apac_2": return .APAC_2
        case "ca" : return .CA
        default:
            return nil
        }
    }
    
    
    @objc(checkForMessages:rejecter:)
    func checkForMessages(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject:@escaping RCTPromiseRejectBlock) -> Void {
        guard let pingOneWalletHelper = self.pingOneWalletHelper else {
            reject("error", "PingOneWalletClient not initialized", nil)
            return
        }
        pingOneWalletHelper.checkForMessages()
        resolve("Message checking initiated")
    }
    
    @objc(processRequest:resolver:rejecter:)
    func processRequest(_ request: String, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) -> Void {
        guard let pingOneWalletHelper = self.pingOneWalletHelper else {
            reject("error", "PingOneWalletClient not initialized", nil)
            return
        }
        
        pingOneWalletHelper.processPingOneRequest(request: request, resolver: resolve, rejector:  reject)
    }
    
    @objc
    func getCredentialsList(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) -> Void {
        guard let pingOneWalletHelper = self.pingOneWalletHelper else {
            reject("error", "PingOneWalletClient not initialized", nil)
            return
        }
        
        pingOneWalletHelper.getCredentialsList(resolver: resolve)
    }
    
    @objc
    func deleteCredential(_ credentialId: String) -> Void {
        guard let pingOneWalletHelper = self.pingOneWalletHelper else {
            return
        }
        
        pingOneWalletHelper.deleteCredential(credentialId: credentialId)
    }
    
    @objc(readNotifications:rejecter:)
    func readNotifications(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) -> Void {
        guard let pingOneWalletHelper = self.pingOneWalletHelper else {
            reject("error", "PingOneWalletClient not initialized", nil)
            return
        }
        
        pingOneWalletHelper.readNotifications(resolver: resolve, rejecter: reject)
    }
    
    @objc
    func constantsToExport() -> [String: Any]! {
        return ["PAIR_WALLET_APPROVE": "approve_pairing",
                "PAIR_WALLET_DENY": "deny_pairing",
                "PRESENTATION_REQUEST_APPROVE": "approve_presentation",
                "PRESENTATION_REQUEST_DENY": "deny_presentation",
                "PAIRING_COMPLETE": "pairing_complete",
                "PRESENTATION_COMPLETE": "presentation_complete"
        ]
    }
    
    @objc
    func enablePolling(){
        guard let pingOneWalletHelper = self.pingOneWalletHelper else {
            return
        }
        pingOneWalletHelper.pollForMessages()
    }
    @objc
    func stopPolling(){
        guard let pingOneWalletHelper = self.pingOneWalletHelper else {
            return
        }
        pingOneWalletHelper.stopPolling()
    }
    
    
}
